core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -0,0 +1,239 @@
+                            return;
+                        }
+                        exception = cause;
+                    } else {
+                        rootReqDuplicator.close();
+                        rootResponse.respond(aRes);
+                        return;
+                    }
+                    retry0(currentAttemptNo, backoff, ctx, rootReqDuplicator, action,
+                           rootResponse, exception, -1);
+                    }));
+            }
+        }));
+    }
+
+    private long getRetryAfterMillis(HttpResponse res) {
+        final HttpHeaders headers = getHttpHeaders(res);
+        long millisAfter = -1;
+        String value = headers.get(HttpHeaderNames.RETRY_AFTER);
+        if (value != null) {
+            try {
+                millisAfter = Duration.ofSeconds(Integer.parseInt(value)).toMillis();
+                return millisAfter;
+            } catch (Exception ignored) {
+                // Not a second value.
+            }
+            try {
+                final ZonedDateTime later =
+                        ZonedDateTime.from(DateTimeFormatter.RFC_1123_DATE_TIME.parse(value));
+                final ZonedDateTime current = ZonedDateTime.now(later.getZone());
+                millisAfter = ChronoUnit.MILLIS.between(current, later);
+            } catch (Exception ignored) {
+                logger.debug("The retryAfter: {}, from the server is neither a http date nor a second.", value);
+            }
+        }
+        return millisAfter;
+    }
+
+    private HttpHeaders getHttpHeaders(HttpResponse res) {
+        final CompletableFuture<HttpHeaders> future = new CompletableFuture<>();
+        final HeaderSubscriber subscriber = new HeaderSubscriber(future);
+        res.closeFuture().whenComplete(subscriber);
+        res.subscribe(subscriber);
+        return future.join(); // Neither blocks here nor throws an exception because it already has headers.
+    }
+
+    private void retry0(int currentAttemptNo, Backoff backoff, ClientRequestContext ctx,
+                        HttpRequestDuplicator rootReqDuplicator, Function<HttpRequest, HttpResponse> action,
+                        DefaultHttpResponse rootResponse, Throwable exception, long millisAfter) {
+        long nextInterval = backoff.nextIntervalMillis(currentAttemptNo);
+        if (nextInterval < 0) {
+            rootResponse.close(exception);
+        } else {
+            final EventLoop eventLoop = ctx.contextAwareEventLoop();
+            nextInterval = Math.max(nextInterval, millisAfter);
+            if (nextInterval <= 0) {
+                eventLoop.submit(() -> retry(currentAttemptNo + 1, backoff, ctx,
+                                             rootReqDuplicator, action,
+                                             rootResponse));
+            } else {
+                eventLoop.schedule(
+                        () -> retry(currentAttemptNo + 1, backoff, ctx,
```

Comment:
Yeap. You're right. I probably need the value that limits the `nextInterval` so that the eventLoop is not scheduled far behind. I think the `idleTimeout` value from `SessionOptions` is proper, but I don't see how to get it, right now.